### PR TITLE
Add data source for regions

### PR DESCRIPTION
**Pull request by syntaqx**
First draft for being able to list regions for an account. Not sure if I'm happy with the usage but it's the best I could think of for the time being. Would love input.

```hcl
data "digitalocean_regions" "foobar" {
}

output "digitalocean_regions" {
   value = digitalocean_regions.foobar.regions
}

# regions = ["nyc1", "sfo1", "nyc2", ...]
```

In reference to #263

> __Note:__ The update to the `vendor/github.com/mitchellh/copystructure/README.md` is Git fixing CRLF whitespacing. Let me know if I need to figure out how to undo this - it's automatic for my git setup unfortunately.

### CODE REVIEW

File: digitalocean/datasource_digitalocean_regions.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+	"github.com/digitalocean/godo"
+	"github.com/hashicorp/terraform/helper/hashcode"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func dataSourceDigitalOceanRegions() *schema.Resource {
+	return &schema.Resource{
+		Read:   dataSourceDigitalOceanRegionsRead,
+		Schema: map[string]*schema.Schema{},
```
Comment (andrewsomething) [inline digitalocean/datasource_digitalocean_regions.go]: My initial thought would be for the schema to look something like:
```suggestion
		Schema: map[string]*schema.Schema{
			"regions": {
				Type:     schema.TypeSet,
				Computed: true,
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"name": {
							Type:     schema.TypeString,
							Computed: true,
						},
						"slug": {
							Type:     schema.TypeString,
							Computed: true,
						},
					},
				},
			},
		},
```

Which makes the UI like:

```
data "digitalocean_regions" "foobar" {
}

resource "digitalocean_droplet" "web" {
  image  = "ubuntu-18-04-x64"
  name   = "web-1"
  region = "nyc2"
  size   = digitalocean_regions.foobar.regions[0].slug
}

output "digitalocean_regions" {
   value = digitalocean_regions.foobar.regions
}

# digitalocean_regions = [
#   {
#     "name" = "New York 1"
#     "slug" = "nyc1"
#   },
#   {
#     "name" = "New York 2"
#     "slug" = "nyc2"
#   },
#  # and so on...
# ]
```

I think this matches the API more closely and makes it more extensible for any additional information about the regions that might be useful.

**Comment by andrewsomething** [comment]
Hi @syntaqx! Sorry for the delayed review. I think this is a great start, and it will definitely be very useful. I've made a suggestion inline about a different approach for the schema. Let me know what you think.

**Comment by syntaqx** [comment]
@andrewsomething If we're going to go that far, might as well just expose the whole API object. Let me know if you like agree!

**Comment by andrewsomething** [comment]
> If we're going to go that far, might as well just expose the whole API object. Let me know if you like/agree!

Sounds like a plan!

File: digitalocean/datasource_digitalocean_regions.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		page, err := resp.Links.CurrentPage()
+		if err != nil {
+			return fmt.Errorf("Error retrieving ssh keys: %s", err)
+		}
+
+		opts.Page = page + 1
+	}
+
+	d.SetId(fmt.Sprintf("%d", hashcode.String(fmt.Sprintf("%+v\n", opts))))
+	d.Set("regions", regionList)
```
Comment (andrewsomething) [inline digitalocean/datasource_digitalocean_regions.go]: For this kind of nested schema, you'll need to loop through the results with something like:

```
	regionsResult := make([]map[string]interface{}, 0)
	for _, r := range regionList {
		item := make(map[string]interface{})
		item["name"] = r.Name
		item["slug"] = r.Slug
		regionsResult = append(regionsResult, item)
	}

	d.Set("regions", regionsResult)
```

Example elsewhere in the code base:
https://github.com/terraform-providers/terraform-provider-digitalocean/blob/121c319d89b21926cb92fe1557c6c88f5e4294ce/digitalocean/resource_digitalocean_database_cluster.go#L334

**Comment by andrewsomething** [comment]
Closing in favor of https://github.com/terraform-providers/terraform-provider-digitalocean/pull/380

**Pull request closed without merging**